soci/ztoc_serializer.go

```diff
@@ -0,0 +1,227 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package soci
+
+import (
+	"bytes"
+	"io"
+	"sort"
+	"time"
+
+	ztoc_flatbuffers "github.com/awslabs/soci-snapshotter/soci/fbs/ztoc"
+	flatbuffers "github.com/google/flatbuffers/go"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// ZtocSerializer incapsulates the serialization logic for Ztoc.
+// Currently, it implements serialization/deserialization to/from flatbuffers.
+type ZtocSerializer struct {
+}
+
+// Serialize serializes Ztoc to its flatbuffers schema and returns a reader along with the descriptor (digest and size only).
+// In case if there's any error situation during serialization to flatbuffers, there will be a panic issued.
+// This is due to the nature of flatbuffers go library, which issues panic for any error.
+func (zs ZtocSerializer) Serialize(ztoc *Ztoc) (io.Reader, ocispec.Descriptor, error) {
+	flatbuf := zs.ztocToFlatbuffer(ztoc)
+	buf := bytes.NewReader(flatbuf)
+	dgst := digest.FromBytes(flatbuf)
+	size := len(flatbuf)
+	return buf, ocispec.Descriptor{
+		Digest: dgst,
+		Size:   int64(size),
+	}, nil
+}
+
+// Deserialize takes the reader with flatbuffers byte stream and deserializes it ztoc.
+// In case if there's any error situation during deserialization from flatbuffers, there will be a panic issued.
```

Comment:
Can we do a `recover()` and issue an error? I can imagine someone would be able to pass in an arbitrary blob as a Ztoc. If that happens, it would crash the snapshotter and all workloads would be impacted.

Reply:
Done. I think, that makes sense.